## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

data = pd.read_csv('Resources/client_dataset.csv')
data.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
column_names = data.columns.tolist()

print(column_names)

['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id', 'order_date', 'order_week', 'order_year', 'item_id', 'category', 'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty', 'line_number']


In [3]:
# Use the describe function to gather some basic statistics
statistics = data.describe()
print(statistics)

          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.
data = data[["client_id", "order_year", "unit_weight"]]
data.head()

client_id  order_year  unit_weight
0      58515        2023         7.50
1      37609        2023         1.49
2      57113        2023         1.68
3      46554        2023         1.23
4      92089        2023        46.43

In [5]:
# What three item categories had the most entries?
data = pd.read_csv('Resources/client_dataset.csv')

top_categories = data['category'].value_counts().head(3)
print(top_categories)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [6]:
# For the category with the most entries, which subcategory had the most entries?
most_common_category = data['category'].value_counts().idxmax()

most_common_subcategory = (data[data['category'] == most_common_category]['subcategory'].value_counts().idxmax())

print(most_common_category, most_common_subcategory)

consumables bathroom supplies


In [7]:
# Which five clients had the most entries in the data?
data = pd.read_csv('Resources/client_dataset.csv')

top_clients = data['client_id'].value_counts().head(5)
print(top_clients)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [8]:
# Store the client ids of those top 5 clients in a list.
top_clients = data['client_id'].value_counts().head(5)

print(top_clients)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [9]:
# How many total units (the qty column) did the client with the most entries order order?
#total_units_ordered = data[data['client_id'] == client_with_most_entries]['qty'].sum()
data = pd.read_csv('Resources/client_dataset.csv')
most_entries_client = data['client_id'].value_counts().idxmax()
filtered_data = data[data['client_id'] == most_entries_client]

total_units_ordered = filtered_data['qty'].sum()
print(total_units_ordered)

64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
data['subtotal'] = data['unit_price'] * data['qty']
print(data.head())

       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor  

In [11]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
data['shipping_price'] = data['unit_weight'].apply(lambda x: 7 * x if x > 50 else 10 * x)
print(data.head())


       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor  

In [12]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%


In [13]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
sales_tax_rate = 0.0925
data['total_price'] = data['subtotal'] + data['shipping_price'] + (data['subtotal'] + data['shipping_price']) * sales_tax_rate
print(data.head())


       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            w

In [14]:
# Create a column for the profit of each line using line cost and line price
data['total_price'] = data['unit_price'] * data['qty']
data['line_cost'] = data['unit_cost'] * data['qty']
data['line_profit'] = data['total_price'] - data['line_cost']
print(data.head())

       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.5

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [15]:
# Check your work using the totals above
order_totals = data.groupby('order_id')['total_price'].sum()
provided_totals = {
    2742071: 152811.89,
    2173913: 162388.71,
    6128929: 923441.25
}
comparison = pd.DataFrame({
    'Calculated Total': order_totals.loc[[2742071, 2173913, 6128929]],
    'Provided Total': pd.Series(provided_totals)
})
comparison['Match'] = comparison['Calculated Total'].round(2) == comparison['Provided Total']
print(comparison)

         Calculated Total  Provided Total  Match
2742071         127799.33       152811.89  False
2173913         109440.76       162388.71  False
6128929         688564.27       923441.25  False


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [16]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
# data = pd.read_csv('Resources/client_dataset.csv')

#for client_id in top_5_client_list:
client_spending = data.groupby('client_id')['total_price'].sum()
top_clients_spending = client_spending.sort_values(ascending=False).head(5)
print(top_clients_spending)

client_id
78965    5.950470e+08
85931    2.071928e+08
38473    1.636205e+08
86721    1.015831e+08
84344    9.287183e+07
Name: total_price, dtype: float64


In [32]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit.
data_dict = {'client_id': top_clients_spending}
columns = ['shipping_price', 'total_price', 'line_cost', 'line_profit']


In [33]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns.
money_df = data[columns]

# Define a function that converts a dollar amount to millions.
def dollars_to_millions(money_df):
    for column in money_df.columns:
        money_df[column] = money_df[column].apply(lambda x: '${:,.2f}'.format(x))    
    
    return money_df


# Apply the currency_format_millions function to only the money columns. 
dfm_d = dollars_to_millions(money_df)
dfm_d 

# Rename the columns to reflect the change in the money format. 




C:\Users\kevin\AppData\Local\Temp\ipykernel_3532\2857855203.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  money_df[column] = money_df[column].apply(lambda x: '${:,.2f}'.format(x))


shipping_price  total_price   line_cost line_profit
0             $75.00  $115,164.00  $80,084.55  $35,079.45
1             $14.90      $523.95     $316.89     $207.06
2             $16.80      $527.28     $306.54     $220.74
3             $12.30    $1,056.18     $720.65     $335.53
4            $464.30    $3,902.00   $2,163.40   $1,738.60
...              ...          ...         ...         ...
54634         $22.50    $2,743.29   $1,702.80   $1,040.49
54635        $117.00    $9,709.73   $8,246.62   $1,463.11
54636         $41.60   $31,188.50  $27,222.25   $3,966.25
54637        $180.40      $165.76     $136.64      $29.12
54638         $20.70    $3,103.31   $1,659.91   $1,443.40

[54639 rows x 4 columns]

In [34]:
# Sort the updated data by "Total Profit (millions)" from highest to lowest and assign the sort to a new DatFrame.
final = dfm_d.sort_values(by="total_price",ascending=False)
final

shipping_price total_price line_cost line_profit
22893         $41.50     $999.95   $645.05     $354.90
39681        $140.90     $999.90   $759.66     $240.24
13800        $140.90     $999.90   $759.66     $240.24
37069          $5.40     $999.39   $743.61     $255.78
53168         $33.40     $999.00   $781.50     $217.50
...              ...         ...       ...         ...
7167          $23.00       $0.00     $0.00       $0.00
39209        $124.20       $0.00     $0.00       $0.00
13609          $5.00       $0.00     $0.00       $0.00
38399          $1.80       $0.00     $0.00       $0.00
4444          $45.10       $0.00     $0.00       $0.00

[54639 rows x 4 columns]